File: Difficulty/Hard/[124] Binary Tree Maximum Path Sum/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        
        def maxPathSumHelper(root):
            nonlocal val
            if root is None: return 0
            left = maxPathSumHelper(root.left)
            right = maxPathSumHelper(root.right)
            lmr = root.val + left + right
            ret = root.val + max(0, max(right, left))
            val = max(val, max(ret, lmr))
            return ret
        val = -2 ** 31 - 1
        maxPathSumHelper(root)
        return val
        

File: Difficulty/Hard/[124] Binary Tree Maximum Path Sum/solution.cpp
/*
 * @lc app=leetcode id=124 lang=cpp
 *
 * [124] Binary Tree Maximum Path Sum
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int val = INT_MIN;
        maxPathSumHelper(root, val);
        return val;
    }

    int maxPathSumHelper(TreeNode* root, int& val){
        if (root == nullptr) return 0;
        int left = maxPathSumHelper(root->left, val);
        int right = maxPathSumHelper(root->right, val);
        int lmr = root->val + left + right;
        int ret = root->val + max(0, max(left, right));        
        val = max(val, max(ret, lmr));
        return ret;
    }
};
// @lc code=end

